wsd-frontend/src/lib/fileUtils.ts

```diff
@@ -0,0 +1,101 @@
+    case 'png':
+      return 'image/png'
+    case 'jfif': // JPEG File Interchange Format - It's "currently" interchangeable with jpeg https://en.wikipedia.org/wiki/JPEG_File_Interchange_Format
+    case 'jpg':
+    case 'jpeg':
+      return 'image/jpeg'
+    case 'webp':
+      return 'image/webp'
+    default:
+      return 'image/png' // Default to PNG if unknown
+  }
+}
+
+export function isImageMimeType(mimeType: string): mimeType is MimeType {
+  return mimeType === 'image/png' || mimeType === 'image/jpeg' || mimeType === 'image/webp'
+}
+
+export function escapeName(name: string): string {
+  // Remove invalid characters for file names
+  return name.replace(/[^a-zA-Z0-9-_.]/g, '_')
+}
+
+function safeFileName(filename: string, maxLength: number = 255) {
+  const extension = filename.slice(filename.lastIndexOf('.'))
+  const name = escapeName(filename.slice(0, filename.lastIndexOf('.')))
+  if (filename.length <= maxLength) return filename
+  return name.slice(0, maxLength - extension.length) + extension
+}
+
+/***
+ * Downloads an image from a given URL and formats it to a specified MIME type.
+ *
+ * If the given mimeType is not supported by the browser, it defaults to PNG.
+ * @param src - The source URL of the image to download.
+ * @param name - The name to save the image as. Default is 'image.png'.
+ * @param mimeType - The MIME type to format the image to. The default is 'image/png'. See https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toBlob#browser_compatibility for compatibility with browsers.
+ * @param quality - The quality of the image (0-1) for JPEG and WEBP formats. Ignored for PNG.
+ */
+export async function downloadAndFormatImage(
+  src: string,
+  name: string = 'image.png',
+  mimeType: MimeType = guessImageMimeType(src),
+  quality: number = 1
+) {
+  name = safeFileName(name)
+  try {
+    // Create an Image element to load the image
+    const img = new Image()
+    img.crossOrigin = 'anonymous' // Handle CORS
+
+    await new Promise((resolve, reject) => {
+      img.onload = resolve
+      img.onerror = reject
+      img.src = src
+    })
+
+    // Create canvas and context
+    const canvas = document.createElement('canvas')
+    canvas.width = img.width
+    canvas.height = img.height
+    const ctx = canvas.getContext('2d')
+
+    if (!ctx) {
+      throw new Error('Could not get canvas context')
+    }
+
+    // Draw image to canvas
+    ctx.drawImage(img, 0, 0)
+
+    // Convert to PNG
+    canvas.toBlob(
+      (blob) => {
+        if (!blob) {
+          throw new Error('Could not generate blob')
+        }
+        // Create download URL and trigger download
+        const link = document.createElement('a')
+        const downloadUrl = URL.createObjectURL(blob)
+        try {
+          link.href = downloadUrl
+          link.download = name
+          document.body.appendChild(link)
+          link.click()
+        } finally {
+          // Revoke the object URL to free up memory
```

Comment:
This one is a good one imo, explains something that's not explicitly obvious from the code